gh-130895: fix multiprocessing.Process join/wait/poll races

duaneg opened the pull request:
This bug is caused by race conditions in the poll implementations (which are called by join/wait) where if multiple threads try to reap the dead process only one "wins" and gets the exit code, while the others get an error.

In the forkserver implementation the losing thread(s) set the code to an error, possibly overwriting the correct code set by the winning thread. This is relatively easy to fix: we can just take a lock before waiting for the process, since at that point we know the call should not block.

In the fork and spawn implementations the losers of the race return before the exit code is set, meaning the process may still report itself as alive after join returns. Fixing this is trickier as we have to support a mixture of blocking and non-blocking calls to poll, and we cannot have the latter waiting to take a lock held by the former.

The approach taken is to split the blocking and non-blocking call variants. The non-blocking variant does its work with the lock held: since it won't block this should be safe. The blocking variant releases the lock before making the blocking operating system call. It then retakes the lock and either sets the code if it wins or waits for a potentially racing thread to do so otherwise.

If a non-blocking call is racing with the unlocked part of a blocking call it may still "lose" the race, and return None instead of the exit code, even though the process is dead. However, as the process could be alive at the time the call is made but die immediately afterwards, this situation should already be handled by correctly written code.

To verify the behaviour a test is added which reliably triggers failures for all three implementations. A work-around for this bug in a test added for gh-128041 is also reverted.

<!-- gh-issue-number: gh-130895 -->
* Issue: gh-130895
<!-- /gh-issue-number -->

duaneg committed:
gh-130895: fix multiprocessing.Process join/wait/poll races

This bug is caused by race conditions in the poll implementations (which are
called by join/wait) where if multiple threads try to reap the dead process
only one "wins" and gets the exit code, while the others get an error.

In the forkserver implementation the losing thread(s) set the code to an error,
possibly overwriting the correct code set by the winning thread. This is
relatively easy to fix: we can just take a lock before waiting for the process,
since at that point we know the call should not block.

In the fork and spawn implementations the losers of the race return before the
exit code is set, meaning the process may still report itself as alive after
join returns. Fixing this is trickier as we have to support a mixture of
blocking and non-blocking calls to poll, and we cannot have the latter waiting
to take a lock held by the former.

The approach taken is to split the blocking and non-blocking call variants. The
non-blocking variant does its work with the lock held: since it won't block
this should be safe. The blocking variant releases the lock before making the
blocking operating system call. It then retakes the lock and either sets the
code if it wins or waits for a potentially racing thread to do so otherwise.

If a non-blocking call is racing with the unlocked part of a blocking call it
may still "lose" the race, and return None instead of the exit code, even
though the process is dead. However, as the process could be alive at the time
the call is made but die immediately afterwards, this situation should already
be handled by correctly written code.

To verify the behaviour a test is added which reliably triggers failures for
all three implementations. A work-around for this bug in a test added for
gh-128041 is also reverted.

duaneg committed:
Add blurb

zmedico commented on Lib/multiprocessing/popen_forkserver.py:
```diff
@@ -64,11 +66,14 @@ def poll(self, flag=os.WNOHANG):
             timeout = 0 if flag == os.WNOHANG else None
             if not wait([self.sentinel], timeout):
                 return None
```
Maybe we should check if self.returncode was set by another thread before we return None here?

duaneg replied:
Hmm, yeah, perhaps so.

If this happens (another thread sets the return code _just_ as the timeout expires) we are racing and it is a coincidence of timing whether `poll` returns `None` or the return code set by the other thread. So I think either is fine, but of course returning the code would be better if we can, so why not?

Actually, we don't need to check it as such, but rather just return it: if it isn't racing it will be `None` anyway.

zmedico commented on Lib/multiprocessing/popen_fork.py:
```diff
@@ -17,23 +18,71 @@ def __init__(self, process_obj):
+            # Child process doesn't exist because it hasn't started yet (see
+            # bpo-1731717) or has already been awaited on a racing thread (see
+            # gh-130895)
+            pass
```
We could hold a threading.Lock during the blocking waitpid call in order to ensure that no more than one thread attempts this at once.

zmedico replied:
Actually, maybe it's fine since we allow the OSError due to the race.

duaneg replied:
We could do so here, but we can't in `_nonblocking_poll`, since that would mean it could block indefinitely. Unfortunately I think that means a lock won't work for our purposes, since it won't help in the case of a blocking `poll` racing against a non-blocking `poll`.

duaneg committed:
Return status code after timeout: usually this will be None, as we are
returning now, except if we raced with another thread that set it just after
our timeout expired.

duaneg committed:
Merge remote-tracking branch 'origin/main' into waiting/gh-130895

duaneg committed:
Ignore fork-in-thread deprecation warnings in test, as now required

YvesDup commented on Lib/multiprocessing/popen_fork.py:
```diff
@@ -17,23 +18,71 @@ def __init__(self, process_obj):
+
+        with self._exit_condition:
+            self._exit_blockers -= 1
             if pid == self.pid:
```
What do you think about addding the following test, just between the increment and the `pid` test:
```       
            if self.returncode is not None:
                return self.returncode
```
I suggest this change because while the condition `RLock` was released, the `self.returncode` attribute could have been updated.

duaneg replied:
I don't _think_ so, but it is very possible I've missed something, this is all (way too) subtle...

If the thread has won the race and `os.waitpid` returned the exit code then the `returncode` cannot have been set: only one thread will get a valid exit code, so if this one has it no other one can. In that case the `if` would always fail.

If the thread did not win the race then the winner might have set it, but in that case it isn't enough to just exit immediately. We need to notify in case there was another blocked thread which didn't see the `returncode` and is waiting.

I.e. consider the following scenario:
1. Threads 1 & 2 are blocked doing an `os.waitpid` in `poll`.
2. Thread 3 does a non-blocking `poll` which "wins" and gets the status code (_but has not yet set it_!)
3. Thread 1 wakes up. It has lost the race, `returncode` hasn't yet been set, and there is another blocker, so it goes into the `while` loop and waits.
4. Thread 3 sets the `returncode`.
5. Thread 2 wakes up. If we add the `if` it will exit without notifying thread 1, which will then be blocked indefinitely.

YvesDup replied:
> If the thread has won the race and os.waitpid returned the exit code then the returncode cannot have been set: 

Are you describe a case of blocking poll ? 
In case of non-blocking poll, the `self._set_returncode` method is always called in a protected block via the condition RLock. The **set** and **notify** operations are always executed **together** in the `self._set_returncode` method . 

> only one thread will get a valid exit code, so if this one has it no other one can. 

I agree

> In that case the if would always fail.

Yes it would failed, but only for the current thread which will continue to run. And the next threads, should not be failed if `self.return code` is set correctly.

> If the thread did not win the race then the winner might have set it, but in that case it isn't enough to just exit immediately. We need to notify in case there was another blocked thread which didn't see the returncode and is waiting.

IMO the notification was already sent from the last call to the`self._set_returncode` method. Even though I have my doubts, if a new notification is really necessary -  when `self._exit_blockers` is zero ? - the new `if` is no longer a relevant option.

> 2\. Thread 3 does a non-blocking `poll` which "wins" and gets the status code (_but has not yet set it_!)

Regarding my first command, I guess yout talk about a blocking poll.

When I ran your `test_racing_joins` test with the new `if`, all seems okay, I never noticed problem with missing notifications. We can check together if you wish.

Your fix seems okay to me. I was just wondering if it was possible to avoid executing the second protected block entirely. 
Thank you for your time.

duaneg replied:
> > If the thread did not win the race then the winner might have set it, but in that case it isn't enough to just exit immediately. We need to notify in case there was another blocked thread which didn't see the returncode and is waiting.
> 
> IMO the notification was already sent from the last call to the`self._set_returncode` method. Even though I have my doubts, if a new notification is really necessary - when `self._exit_blockers` is zero ? - the new `if` is no longer a relevant option.

You make a good point, and I made a mistake in my analysis above. The early exit you suggest is indeed safe if another racing `poll` has won and set the return code, since they will have also notified. The problematic case is the pathological one, where some other code entirely called `os.waitpid` and won the race. In that case `_set_returncode` is not called and hence we need that backup notification to avoid a hang.

In that scenario the `Popen` object is broken, since it has no way to get the child's exit code. Methods like `is_alive()` and mp mechanisms downstream of `Popen` will give the wrong result (there are a _bunch_ of open issues that boil down to this). It is a "user error" situation, but we should still try and make everything fail as gracefully as possible (i.e. not hang :wink:).

So, we could put the early exit in. Ultimately I think the code will behave correctly with or without it. Personally, I think the code reads slightly simpler without it, just in basic terms of fewer conditionals/less cyclomatic complexity, but YMMV on that. If there is a general consensus otherwise I'll happily add it.

> When I ran your `test_racing_joins` test with the new `if`, all seems okay, I never noticed problem with missing notifications. We can check together if you wish.

Thanks, I would love to collaborate on this! The existing test case does not test the pathological behaviour, so it will indeed miss this. Here is a new test script that does; it hangs reliably for me if the last blocker notification is removed:
```python
import multiprocessing as mp
import os
import threading

N=8

def wait(barrier, p):
    barrier.wait()
    p.join()

    # WRONG due to losing the child's status
    assert p.exitcode is None
    assert p.is_alive()

def race():
    pbarrier = mp.Barrier(2)

    # Ensure child has completed
    p = mp.Process(target=pbarrier.wait, args=(pbarrier,))
    p.start()
    pbarrier.wait()

    # Steal the child's exit code
    pid, sts = os.waitpid(p.pid, 0)
    assert pid == p.pid
    assert sts == 0

    # Since the child is dead, these should not block...
    tbarrier = threading.Barrier(N)
    threads = [threading.Thread(target=wait, args=(tbarrier, p)) for _ in range(N)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

if __name__ == "__main__":
    mp.set_start_method('fork')
    race()
```

Does that reproduce for you? I guess it should be tidied up and turned into a unit test...

> Your fix seems okay to me. I was just wondering if it was possible to avoid executing the second protected block entirely. Thank you for your time.

Thanks! For the record, I very much appreciate your review, and am grateful for your feedback and intelligent questions. I really dislike the sort of overly-complex synchronisation I'm doing here, so I'm particularly grateful for it in this case. I would _love_ to simplify this!

YvesDup replied:
Thank you for your last comments too. I will work on it tomorrow and hope to find a simpler proposal together.

YvesDup replied:
> The problematic case is the pathological one, where some other code entirely called os.waitpid and won the race. In that case _set_returncode is not called and hence we need that backup notification to avoid a hang.
> In that scenario the Popen object is broken, since it has no way to get the child's exit code. Methods like is_alive() and mp mechanisms downstream of Popen will give the wrong result (there are a bunch of open issues that boil down to this). It is a "user error" situation, but we should still try and make everything fail as gracefully as possible (i.e. not hang 😉).

Yes it is an user error but I disagree that we should fix this problem. Mixing  operating system primitives and the multiprocessing module is very hacky and neither appropriate nor clever, even if it is possible. IMO this is a user error on the same level as setting directly 0 to `self.returncode` or calling `_exit_condition.notify_all` from a user code. Are not we ** consenting adults** ?
The opinion of a core dev would be welcome, right?

That said, I haven't found a case where we have to call `notify_all` from the `if self._exit_blokers == 0:` test (located in the second protected block). And that bothers me a lot. Is this part really necessary? I had also run a test forcing the evalution of this test. There was never any pending thread to notify.
Other points: CI are green for test_multiprocessing_spawn/forkserver and test_concurrent_futures - I can not run test_multiprocessing_fork because I am working on MacOSX -

I added logs in the `poll` method. [See here](https://github.com/YvesDup/cpython/blob/6128f60d7505d043bd9a78287138972481b6f5bc/Lib/multiprocessing/popen_fork.py#L23) and [here](https://github.com/YvesDup/cpython/blob/6128f60d7505d043bd9a78287138972481b6f5bc/Lib/multiprocessing/popen_fork.py#L29-L104) in order to confirm my analysis.
Pehars I missed somthing.

If you run the unit test with my logs, you can print the logs after the `join` as below:
```
p = mp.Process(....)
...
...
    t.join() # t is a thread
    print(p._popen._logs)
```
> So we could add the early exit.

I'm +1 about this, because executing the end of the second protected block while `self.returncode` is set, seems to me not necessary.  

I'm sorry I don't have a simpler proposal at this time.

duaneg replied:
> Yes it is an user error but I disagree that we should fix this problem.

I'm not saying we should fix it, indeed we _cannot_. IMO we should, however, fail as gracefully as possible. In particular, we should not hang if we do not need to.

> That said, I haven't found a case where we have to call `notify_all` from the `if self._exit_blokers == 0:` test

Doesn't the test script I posted in the comment above reproduce this for you? On my machine it always hangs, but depending on your machine it may be less reliable: I suggest calling `race` in a loop if it isn't hanging immediately for you.

Or are you saying you cannot reproduce it _only_ calling `poll`, not reaping the child from external code? If so then yes, I agree this hang won't be triggered if the child process is reaped by `poll`.

> I'm sorry I don't have a simpler proposal at this time.

That's fine, I couldn't come up with one myself after all, and perhaps there isn't one. I very much appreciate your effort and contributions to improving the patch, regardless!

YvesDup replied:
> Doesn't the test script I posted in the comment above reproduce this for you? On my machine it always hangs, but depending on your machine it may be less reliable: I suggest calling `race` in a loop if it isn't hanging immediately for you.

I can not reproduce the hang, I removed my logs, the early exit and add a loop in `race`. It never hangs,  there is something weird. I tried with Ubuntu with the same result.

duaneg replied:
> I can not reproduce the hang, I removed my logs, the early exit and add a loop in `race`. It never hangs, there is something weird. I tried with Ubuntu with the same result.

Did you remove the notify-on-last-blocker? The test is intended to demonstrate that is required<sup>1</sup>: with it it will complete, without it it will hang. The early exit doesn't make any _significant_<sup>2</sup> difference either way, which is why I would prefer not to add it. IMO the code is simpler without it, so we shouldn't add it without reason, and there isn't a good one.

Of course, if I'm missing a reason why it matters, please let me know!

<sup>1</sup> At least, to avoid a significantly worse debugging experience for users who run into this issue.
<sup>2</sup> It does avoid an unnecessary notify-on-last-blocker-completing, but this is not hot code and a notify with no waiters<sup>3</sup> is essentially a no-op, so it has no meaningful performance impact.
<sup>3</sup> As must be the case, given this is the last completing blocker and `returncode` has been set.

YvesDup replied:
> Did you remove the notify-on-last-blocker?

~~I ran this test with your initial proposal. Removing the **notify-on-last-blocker** has no effect on the result: no hang. This really bothers me.
As I write this comment, I wonder how is this `race` test with the `spawn` startup method, since “fork” and “threading” are not really friredly.~~

Update: I started from your last commit, and I am relieved to announce that the `race` test hangs when the  **notify-on-last-blocker** is removed, in `fork` and `spawn` start methods. Sorry for the noise.

I understand your position about **early exit** (not add it) regarding the useful of **notify-on-last-blocker** to avoid hangs. 
Thank you for this conversation

duaneg replied:
No bother at all, I feel more confident about the code after digging in with you, so it has been a useful exercise!

I tried to turn the reproduction script above into a unit test, but it gets rather tricky, and at present at least I don't think it is worth the effort/complexity. If we ever look at revising how we deal with externally-reaped-children on a wider level we could revisit it then.

YvesDup commented on Lib/multiprocessing/popen_fork.py:
```diff
@@ -17,23 +18,71 @@ def __init__(self, process_obj):
+            # racing calls that might be about to set it
+            while self.returncode is None and self._exit_blockers > 0:
+                self._exit_condition.wait()
+
```
I added the following statement here  `assert self.returncode is not None` and ran your test (and others from related issues). And if I haven't missed anything, at the exit of the `while` loop, the `self.returncode` attr is never set to `None`. 
So I wonder if it would be consistent to remove `self._exit_blockers > 0`  from the `while` loop test ?

You could also replace this loop with a `Condition.wait_for` method. It's up to you.

duaneg replied:
Unfortunately it is possible for `returncode` to be `None`: some other code entirely might be doing an `os.waitpid` and hence the exit code is effectively "lost". I.e. something like this:

```python
import multiprocessing as mp
import os
import threading

def race():
    mp.set_start_method('fork')
    barrier = mp.Barrier(2)

    p = mp.Process(target=barrier.wait, args=(barrier,))
    p.start()
    barrier.wait()

    # Wait for the process here, bypassing the mp machinery...
    pid, sts = os.waitpid(p.pid, 0)
    assert pid == p.pid
    assert sts == 0

    # This will fail since the exit code was "lost" to the waitpid above
    p.join()
    assert p.exitcode == 0
    assert not p.is_alive()

if __name__ == "__main__":
    race()
```

duaneg replied:
Thanks for mentioning `Condition.wait_for`, I hadn't noticed that. I would normally prefer it to an explicit loop, but in this case I think it makes the code slightly more verbose/uglier, since we would need to define a function or lambda to pass in as the predicate.

YvesDup replied:
I agree, `returncode`could be `None`.

YvesDup commented on Lib/multiprocessing/popen_forkserver.py:
```diff
@@ -32,6 +33,7 @@ class Popen(popen_fork.Popen):
 
     def __init__(self, process_obj):
         self._fds = []
+        self._lock = threading.Lock()
```
What are we do here about the private `_exit_condition` and `_exit_blockers` attributes inherit from `popen_fork.Popen` class ? I suggest a private method in the parent class which define its own synchronize attributes, calls from the `__init__`.

duaneg replied:
Good question. I had intended to just ignore them. We are very fortunate the locking is entirely self-contained within each class's `poll` implementation, so this works just fine, but it does leave the child with unused instance variables inherited from the parent.

If I understand your suggestion (please correct me if not), you are suggesting a polymorphic method to initialise locking, so each class only creates the attributes it requires. I agree, that sounds like an improvement, thanks! I'll update the patch shortly to add it.

BTW, I think all this would be simpler and more robust if `popen_forkserver.Popen` did not directly inherit from `popen_fork.Popen`. They should probably both inherit from an abstract base class, instead. Presumably it would be impractical, if not impossible, to change that now due to backward compat concerns, though.

YvesDup replied:
> ... but it does leave the child with unused instance variables inherited from the parent
If I understand your suggestion (please correct me if not), you are suggesting a polymorphic method to initialise locking, so each class only creates the attributes it requires. I agree, that sounds like an improvement, thanks! I'll update the patch shortly to add it.

Thank your for the patch which looks good to me..

> BTW, I think all this would be simpler and more robust if popen_forkserver.Popen did not directly inherit from popen_fork.Popen. They should probably both inherit from an abstract base class, instead. Presumably it would be impractical, if not impossible, to change that now due to backward compat concerns, though.

Peharps to propose in a next refactor issue ?

duaneg replied:
Perhaps!

It is part of the public API, so by my understanding would need to go through the deprecation process (or get an SC exemption).

I can't think of why sensible user code would do anything that would break (e.g. use `isinstance` on a `Popen` object), but of course that doesn't mean much. "Sensible" is subjective and in my experience there is always someone somewhere relying on even the most obscure implementation details for some clever (or "clever") solution.

I guess the thing to do would be _tentatively_ propose it and see what more experienced core devs/maintainers think. I've added it to my notes for the future :wink:.

duaneg committed:
Initialise synchronisation-related attributes in polymorphic method so each
class can do so as it requires.

Co-authored-by: Duprat <yduprat@gmail.com>

gpshead committed:
Merge branch 'main' into gh-130895

gpshead committed:
Merge branch 'main' into gh-130895
